### PR TITLE
Cirrus CI: Prepare for druntime/Phobos CI

### PR DESCRIPTION
**Pull request by kinke**

### CODE REVIEW

**Comment by kinke** [comment]
Alright, enabling Cirrus CI for druntime and Phobos after this should be a matter of adding some ~80 YAML lines to both repos, very similar to how Azure CI is handled.

**Review by WalterBright — CHANGES_REQUESTED** [review]
1. Non-trivial changes should come with at least some explanation of what the PR does other than "Prepare".

2. Bash code is code and should be commented as appropriate.

**Comment by kinke** [comment]
1. See https://github.com/dlang/dmd/pull/11868#issuecomment-707918518. I assumed the reviewer would be somewhat familiar with the CI process, as Azure is set up similarly (but way more complex over all). - The point is to make the YAML files as compact as possible, because they are required for each repo, e.g., by extracting the prerequisites installation to external `cirrusci.sh`. The druntime and Phobos repos clone the dmd repo anyway and can re-use that file for their CI, as well as `ci.sh`, which is also used by Semaphore CI.

2. I assume you mean a little header for new `cirrusci.sh`; if not, please be more specific.

File: ci.sh [hunk 1]
```diff
@@ -93,6 +98,14 @@ test_dmd() {
     fi
 }
 
+test_druntime() {
```
Comment (WalterBright) [inline ci.sh:101]: # run druntime's unit tests

File: ci.sh [hunk 2]
```diff
@@ -93,6 +98,14 @@ test_dmd() {
...
+    make -j$N -C ../druntime -f posix.mak MODEL=$MODEL unittest
+}
+
+test_phobos() {
```
Comment (WalterBright) [inline ci.sh:105]: Functions need a comment.

File: ci.sh [hunk 3]
```diff
@@ -163,22 +180,41 @@ download_install_sh() {
 }
 
 install_d() {
```
Comment (WalterBright) [inline ci.sh:182]: what does this function do?

File: ci.sh [hunk 4]
```diff
@@ -163,22 +180,41 @@ download_install_sh() {
...
+    local gdc_version="${HOST_DC:5}"
+    if [ ! -e ~/dlang/gdc-$gdc_version/activate ] ; then
         sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
         sudo apt-get update
-        sudo apt-get install -y gdc-${GDC_VERSION}
+        sudo apt-get install -y gdc-$gdc_version
         # fetch the dmd-like wrapper
-        sudo wget https://raw.githubusercontent.com/D-Programming-GDC/GDMD/master/dmd-script -O /usr/bin/gdmd-${GDC_VERSION}
-        sudo chmod +x /usr/bin/gdmd-${GDC_VERSION}
+        sudo curl -fsSL -A "$CURL_USER_AGENT" --connect-timeout 5 --speed-time 30 --speed-limit 1024 --retry 5 --retry-delay 5 https://raw.githubusercontent.com/D-Programming-GDC/GDMD/master/dmd-script -o /usr/bin/gdmd-$gdc_version
```
Comment (WalterBright) [inline ci.sh:190]: What's a "dmd-like wrapper" ?

File: ci.sh [hunk 5]
```diff
@@ -163,22 +180,41 @@ download_install_sh() {
...
+        sudo chmod +x /usr/bin/gdmd-$gdc_version
```
Comment (WalterBright) [inline ci.sh:191]: ? I see only one script being created in the code

**Comment by WalterBright** [comment]
Comments are better now. Pls fix the remaining comments then merge. Thanks!

**Review by WalterBright — APPROVED** [review]

File: ci.sh (hunk 5 above)
Comment (kinke) [reply ci.sh:194]: This `gdmd` script is downloaded one line above.

**Comment by kinke** [comment]
Okay, I think your remaining comments (the majority of which not related to changes in this PR :]) should have been addressed.

**Comment by kinke** [comment]
The `std.concurrency` unittest segfault for the one Ubuntu x64 job is unrelated and hopefully no random master regression caused by https://github.com/dlang/phobos/pull/7655:
```
generated/linux/debug/64/unittest/libphobos2-ut.so(_D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPSQCk3sys5posix6signal9siginfo_tPvZv+0x55)[0x7fb2f4d5577d]
/lib/x86_64-linux-gnu/libc.so.6(+0x354c0)[0x7fb2f0dff4c0]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D3std11concurrency5yieldFNbZv+0x4a)[0x7fb2f35fcd42]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D3std11concurrency10MessageBox__T3getTDFNaNbNiNfiZvTPFNaNiNfCQCiQCh14LinkTerminatedZvTPFNaNiNfCQDqQDp15OwnerTerminatedZvTPFNfSQEv7variant__T8VariantNVmi32ZQpZvZQExMFMQEyMQEoMQDkMQCfZb+0x7b)[0x7fb2f360ef6b]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D3std11concurrency__T11receiveOnlyTiZQqFZi+0x88)[0x7fb2f360edbc]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D3std11concurrency19__unittest_L1854_C9FZ13testSchedulerFCQCfQCe9SchedulerZ9__lambda2MFZ9__lambda1FZ12__dgliteral1MFNaNiNfZv+0x11)[0x7fb2f35fcfd9]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D3std9exception__T15assertNotThrownHTC4coreQBm11AssertErrorTvZQBsFNaLvAyaQdmZv+0x35)[0x7fb2f361e1ad]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D3std11concurrency19__unittest_L1854_C9FZ13testSchedulerFCQCfQCe9SchedulerZ9__lambda2MFZ9__lambda1FZv+0x48)[0x7fb2f35fcf7c]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D3std11concurrency__T6_spawnTPFZvZQnFbQjZ4execMFZv+0x40)[0x7fb2f360dc98]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D4core6thread7context8Callable6opCallMFZv+0x41)[0x7fb2f4d6ae71]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D4core6thread10threadbase10ThreadBase3runMFZv+0x19)[0x7fb2f4d67c15]
generated/linux/debug/64/unittest/libphobos2-ut.so(_D4core6thread8osthread6Thread3runMFZv+0x15)[0x7fb2f4d69549]
generated/linux/debug/64/unittest/libphobos2-ut.so(thread_entryPoint+0xe5)[0x7fb2f4d6a65d]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba)[0x7fb2f09b06ba]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7fb2f0ed14dd]
posix.mak:413: recipe for target 'unittest/std/concurrency.run' failed
make[1]: *** [unittest/std/concurrency.run] Segmentation fault (core dumped)
```

**Comment by kinke** [comment]
Ping.

**Comment by kinke** [comment]
This won't automerge due to an autotester 'break'.

**Comment by thewilsonator** [comment]
Really? everything looks green.

**Pull request merged**